crates/rmf_site_editor/src/site/robot_properties.rs

```diff
@@ -28,78 +31,323 @@ pub struct UpdateRobotPropertyKinds {
-    mut removals: RemovedComponents<ModelProperty<Robot>>,
-    mut update_robot_property_kinds: EventWriter<UpdateRobotPropertyKinds>,
-) {
-    let property_label = T::label();
-
-    // Remove Robot property entirely
-    for description_entity in removals.read() {
-        commands.entity(description_entity).remove::<T>();
-        update_robot_property_kinds.write(UpdateRobotPropertyKinds {
-            entity: description_entity,
-            label: property_label.clone(),
-            value: serde_json::Value::Object(Map::new()),
-        });
+#[derive(Default)]
+pub struct RobotPropertiesPlugin {}
+
+impl Plugin for RobotPropertiesPlugin {
+    fn build(&self, app: &mut App) {
+        // Allows us to toggle Robot as a configurable property
+        // from the model description inspector
+        let component_id = app.world_mut().register_component::<ModelProperty<Robot>>();
+        let mut model_property_data = ModelPropertyData::from_world(app.world_mut());
+        model_property_data.optional.insert(
+            component_id,
+            (
+                "Robot".to_string(),
+                |mut e_cmd| {
+                    e_cmd.insert(ModelProperty::<Robot>::default());
+                },
+                |mut e_cmd| {
+                    e_cmd.remove::<ModelProperty<Robot>>();
+                },
+            ),
+        );
+        app.insert_resource(model_property_data);
+        app.world_mut().init_resource::<RobotPropertyRegistry>();
+        app.add_event::<UpdateRobotPropertyKinds>()
+            .add_systems(PostUpdate, update_model_instances::<Robot>);
     }
+}
+
+type InsertDefaultValueFn = fn() -> Result<Value, Error>;
+
+pub struct RobotPropertyRegistration {
+    pub widget: Option<Entity>,
+    pub kinds: HashMap<String, RobotPropertyKindRegistration>,
+}
+
+pub struct RobotPropertyKindRegistration {
+    pub default: InsertDefaultValueFn,
+}
+
+/// This resource keeps track of all the properties that can be configured for a robot.
+#[derive(Resource, Deref, DerefMut)]
+pub struct RobotPropertyRegistry(pub HashMap<String, RobotPropertyRegistration>);
+
+impl FromWorld for RobotPropertyRegistry {
+    fn from_world(_world: &mut World) -> Self {
+        Self(HashMap::new())
+    }
+}
+
+/// Implement this plugin to add a new configurable robot property to the site
+pub struct RobotPropertyPlugin<Property, RecallProperty>
+where
+    Property: RobotProperty,
+    RecallProperty: Recall + Component<Mutability = Mutable> + Default,
+    RecallProperty::Source: RobotProperty,
+{
+    _ignore: std::marker::PhantomData<(Property, RecallProperty)>,
+}
+
+impl<Property, RecallProperty> Default for RobotPropertyPlugin<Property, RecallProperty>
+where
+    Property: RobotProperty,
+    RecallProperty: Recall + Component<Mutability = Mutable> + Default,
+    RecallProperty::Source: RobotProperty,
+{
+    fn default() -> Self {
+        Self {
+            _ignore: Default::default(),
+        }
+    }
+}
```

Comment:
Nit, is it possible to jut derive `Default` and get rid of this custom implementation since it really just calls `default` for the only member variable in the struct?